REVM cheatcodes

onbjerg opened the pull request:
**In draft while I address the remaining items**

Ports all the cheatcodes to REVM in a cheatcode-specific inspector. The cheatcodes have been split up into different modules so we don't end up with a huge file like we had with Sputnik, because we can.

This PR also incldues an update to ethers, which we needed to bump `primitive-types` (needed by `revm`), but it also brought in a lot of other changes. I tried to resolve the breaking changes as best I could, but I'm sure it was changed differently on master.

Also adds `InspectorStack` which is an inspector that calls a bunch of underlying inspectors.

Using my branch of revm until https://github.com/bluealloy/revm/pull/69 is merged.

Remaining items:

- [x] `expectEmit`
- [x] `expectCall`
- [x] Block hashes for `roll`
- [x] Moving to static dispatch (probably?) for `InspectorStack` using an enum
- [ ] Porting over some Rust tests - I already tested all of these changes with https://github.com/onbjerg/foundry-test

Some questions:

- ~~In our `record`/`accesses` test we assert that 2 reads have taken place because "SSTORE does an SLOAD", but this is not reproduced in REVM. I couldn't find any docs that suggest this is supposed to be the case? If it is, or if we want to replicate the behavior, the change is small, but I'd like some thoughts on it~~ (we're now recording a read on every write)
- ~~The `getCode` cheatcode sort of works, but when I added more files to https://github.com/onbjerg/foundry-test the test started failing. As far as I can see, the bytecode does match except for some Solidity metadata at the end. I cannot figure out if this is an error on my part, or if it is because ethers was updated and that somehow changed the metadata that `solc` appends at the end of the bytecode?~~ (seems to work after rebase)

Closes #752 
Closes #755 
Closes #756

onbjerg commented:
Side note, I highly recommend enabling the new Github PR file tree if you are part of the feature preview :smile:

onbjerg commented on forge/src/executor/inspector/cheatcodes/mod.rs:
```diff
@@ -0,0 +1,230 @@
+
+        // Handle expected reverts
+        if let Some(expected_revert) = &self.expected_revert {
+            if data.subroutine.depth() <= expected_revert.depth {
```
This depth check is needed to replicate a specific behavior we have on master now, see this test: https://github.com/onbjerg/foundry-test/blob/7909fe8db74ca13fd86cc501fbbb53baaaec5db8/src/cheats/ExpectRevert.t.sol#L79-L84

The reason this was handled in Sputnik without recording the `depth` is because we wrapped the entirety of `call_inner`

onbjerg commented on forge/src/executor/inspector/cheatcodes/mod.rs:
```diff
@@ -0,0 +1,230 @@
+                        .or_insert_with(Vec::new)
+                        .push(key);
+                }
+                opcode::SSTORE => {
+                    let key = try_or_continue!(interpreter.stack().peek(0));
+                    storage_accesses
+                        .writes
+                        .entry(interpreter.contract().address)
+                        .or_insert_with(Vec::new)
+                        .push(key);
+                }
```
As mentioned in the PR, if we want to replicate the "SLOAD per SSTORE" behavior we just write to `storage_accesses.reads` here as well

gakonst replied:
@brockelmore up to you here, wdyt?

brockelmore replied:
hmmm ideally we do, because it maintains simplicity for storage overriding based on function calls. otherwise in stdstorage, we would have to also check writes

onbjerg replied:
I'll replicate the behaviour in the port :)

onbjerg commented on forge/src/executor/inspector/cheatcodes/env.rs:
```diff
@@ -0,0 +1,171 @@
+            Ok(Bytes::new())
+        }
+        HEVMCalls::Store(inner) => {
+            // TODO: Does this increase gas usage?
```
I still haven't figured this out, will ping Dragan unless one of you know - whenever we want to change something about an account (nonce, storage, balance, ...) we need to call `load_account` first in the subroutine and it does a bunch of stuff. As far as I could tell there is no gas recording in there, it just returns some info about whether the account/slot/whatever was hot or cold, which I assume the rest of the REVM impl uses to account for gas.

Generally, though, there are some major issues with the current REVM port in terms of gas accounting, which I will address in a separate PR

gakonst replied:
Sounds good - maybe @draganrakita can share some context here for anyone following the PR

onbjerg commented on cli/src/cmd/bind.rs:
```diff
@@ -129,7 +129,7 @@ impl BindArgs {
     /// Generate the bindings
     fn generate_bindings(&self) -> eyre::Result<()> {
         let bindings = self.get_multi()?.build()?;
-        println!("Generating bindings for {} contracts", bindings.len());
```
Seems like `bindings.len()` or equiv no longer exists :thinking:?

onbjerg replied:
Fixed in rebase

onbjerg commented:
Also @gakonst I considered adding `getNonce`/`setNonce` and `logs` but a few issues:

- Setting the nonce is not currently possible(?) from inside an inspector. I think we should work towards enabling that, but I'm not sure we want to do this now?
- For a `logs` cheatcode we can definitely add it but the UX is going to be gross. We can add it later in a follow up PR

Tried rebasing a couple of times but it was a huge pain on the `revm` branch and I imagine having to rebase this too on top of that would be extra pain. My current plan is to get this and #835 merged and then rebasing to rip the band aid off in one go

gakonst commented on forge/src/executor/inspector/cheatcodes/env.rs:
```diff
@@ -0,0 +1,171 @@
+    pub writes: BTreeMap<Address, Vec<U256>>,
+}
+
+fn record(state: &mut Cheatcodes) {
```
Is this right? Shouldn't it be taking some extra argument to record? Else, maybe should be called "start record" or something?

onbjerg replied:
Takes no arguments, but we can rename to start record:

https://github.com/gakonst/foundry/blob/a00c1939becf3b40f1453e330d3d9507a756d610/evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs#L812-L815

gakonst commented on forge/src/executor/inspector/cheatcodes/env.rs:
```diff
@@ -0,0 +1,171 @@
+        }
+        HEVMCalls::Roll(inner) => {
+            data.env.block.number = inner.0;
+            // TODO: Set blockhash somehow
```
Let's set it as the hash of the block number? Something to consider here is that this would break when using forking tests and roll (the hash wouldn't be the one from mainnet). I don't think we need to do anything about this, but maybe worth documenting

onbjerg replied:
Yeah, the main reason I've delayed it until now is that it requires us to create our own database or upstream a fix in revm (which I am going to do) since inspectors have no control over it

gakonst commented on forge/src/executor/inspector/cheatcodes/env.rs:
```diff
@@ -0,0 +1,171 @@
+        }
+        HEVMCalls::Load(inner) => {
+            // TODO: Does this increase gas usage?
+            data.subroutine.load_account(inner.0, data.db);
```
Is it possible the account is already loaded in the subroutine?

onbjerg replied:
Yes

gakonst commented on forge/src/executor/inspector/cheatcodes/expect.rs:
```diff
@@ -0,0 +1,203 @@
+
+/// For some cheatcodes we may internally change the status of the call, i.e. in `expectRevert`.
+/// Solidity will see a successful call and attempt to decode the return data. Therefore, we need
+/// to populate the return with dummy bytes so the decode doesn't fail.
```
could  also add comment that 320 bytes was arbitrarily chosen because it's "long enough" for 10 words in the return value of the call being mocked?

gakonst commented on forge/src/executor/inspector/cheatcodes/expect.rs:
```diff
@@ -0,0 +1,203 @@
+        return Err("Call did not revert as expected".to_string().encode().into())
+    }
+
+    if retdata.is_empty() {
```
Should this error only happen if expect_revert is not empty? Thinking about the expectRevert(bytes("")) case

onbjerg replied:
I think this is correct, it's a condensed form of

https://github.com/gakonst/foundry/blob/a00c1939becf3b40f1453e330d3d9507a756d610/evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs#L452-L468

In the case of `expectRevert(bytes(""))` I think there is still data (some `bytes` header)

Would make sense to have some extra eyes on it though

onbjerg replied:
This was a bug but it is fixed now

gakonst commented on forge/src/executor/inspector/cheatcodes/mod.rs:
```diff
@@ -0,0 +1,300 @@
+
+    fn step(&mut self, interpreter: &mut Interpreter, _: &mut EVMData<'_, DB>, _: bool) -> Return {
+        // Record writes and reads if `record` has been called
+        if let Some(storage_accesses) = &mut self.accesses {
```
@brockelmore we should eventually use this slot read/write trace logic for determining the parallel evm dag

gakonst reviewed (commented):
Very nice - so far so good! Hyped. Agree with your assessment of rebasing after we also merge the forked provider

brockelmore commented on forge/src/executor/inspector/cheatcodes/mod.rs:
```diff
@@ -0,0 +1,300 @@
+        if !self
+            .expected_emits
+            .iter()
+            .filter(|expected| expected.depth == data.subroutine.depth())
```
i believe this should actually be a different comparison and is the reason for some of the expectEmit bugs

I have think on it more

brockelmore replied:
disregard wouldn't be here i don't think. should be in the actually emission of the logs

brockelmore commented on forge/src/executor/inspector/cheatcodes/mod.rs:
```diff
@@ -0,0 +1,300 @@
+            .or_else(|| expect::apply(self, data, &decoded))
+            .or_else(|| fuzz::apply(data, &decoded))
+            .or_else(|| ext::apply(self.ffi, &decoded))
+            .ok_or_else(|| "Cheatcode was unhandled. This is a bug.".to_string().encode())?
```
do we know the efficiency loss here by repeatedly trying & failing until we match? i.e. 5 match statements vs 1 match statement? I don't imagine its that bad but just curious if thats a concern here

brockelmore reviewed (commented):
nice :)

onbjerg commented:
Addressed the outstanding items. I moved to static dispatch in 2dd2eb7477243d859f87fa05f3e8d9f35ccebf36 with a helper macro, so I'd like some extra eyes on that. I really liked the dynamic dispatch approach but all the research I could find seemed to suggest that it was much slower than static dispatch

onbjerg commented:
Solmate, this PR:

```
________________________________________________________
Executed in    2,78 secs    fish           external
   usr time   37,90 secs  212,00 micros   37,90 secs
   sys time    0,20 secs   62,00 micros    0,20 secs
```

Solmate, master:

```
________________________________________________________
Executed in    4,95 secs    fish           external
   usr time   51,85 secs  296,00 micros   51,85 secs
   sys time    0,19 secs   85,00 micros    0,19 secs
```

(Prebuilt, no traces or logs)

gakonst commented on forge/src/executor/inspector/logs.rs:
```diff
@@ -96,9 +96,9 @@ where
         if call.contract == *HARDHAT_CONSOLE_ADDRESS {
             let (status, reason) = self.hardhat_log(call.input.to_vec());
-            (status, Gas::new(0), reason)
+            (status, Gas::new(call.gas_limit), reason)
```
@draganrakita it seems like `Gas::new` now needs to be providing the full gas limit per call, vs gas used

gakonst reviewed (approved):
looks great!
